### PR TITLE
[SQL] Support array_append function

### LINKED ISSUE #1414: [SQL] array literal should be Vec<Option<T>> but we get Vec<T>
While implementing `array_append`, for the following test case: 

```java
    @Test
    public void testArrayNull() {
        this.q("""
                SELECT array_append(ARRAY [1, 2], null::int);
                 array_append
                --------------
                 {1, 2, NULL}"""
        );
    }
```

The generated rust code calls the `array_append` function with parameters `Vec<T>` and `None::<i32>` and expects `Vec<Option<T>>`. 

Generated code: 

```rust
            Tup1::new(array_append_vec_N(vec!(
                1i32, 2i32, ), None::<i32>))
```

It should be `vec![ Some(1i32), Some(2i32)]` instead. 

I'm not sure how to cast the inner type to be nullable in the compiler.

### PR DESCRIPTION
**Pull request by mihaibudiu**
Is this a user-visible change (yes/no): yes

Fixes #1414 

@abhizer please add a few more tests

### CODE REVIEW

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/InnerVisitor.java [hunk 1]
```diff
@@ -452,10 +448,18 @@ public VisitDecision preorder(DBSPBlockExpression node) {
         return this.preorder(node.to(DBSPExpression.class));
     }
 
-    public VisitDecision preorder(DBSPApplyExpression node) {
+    public VisitDecision preorder(DBSPApplyBaseExpression node) {
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/InnerVisitor.java:451]: these changes are unrelated

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPApplyBaseExpression.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package org.dbsp.sqlCompiler.ir.expression;
+
+import org.dbsp.sqlCompiler.compiler.errors.InternalCompilerError;
+import org.dbsp.sqlCompiler.compiler.frontend.CalciteObject;
+import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.DBSPTypeAny;
+import org.dbsp.sqlCompiler.ir.type.DBSPTypeFunction;
+
+/** Base class for function applications */
+public abstract class DBSPApplyBaseExpression extends DBSPExpression {
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPApplyBaseExpression.java:10]: this is just a bit of refactoring, not connected to array_append

File: sql-to-dbsp-compiler/lib/sqllib/src/array.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
+// Array operations
+
+use std::ops::Index;
+
+pub fn element<T>(array: Vec<T>) -> Option<T>
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/lib/sqllib/src/array.rs:5]: most of these functions were moved from lib.rs

**Comment by abhizer** [comment]
I think some of the changes I committed were lost when rebasing. Will make those changes, we can merge it then.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java [hunk 4]
```diff
@@ -951,8 +951,31 @@ node, new DBSPTypeBool(CalciteObject.EMPTY, false), DBSPOpcode.EQ,
...
+                    DBSPVariablePath var = new DBSPVariablePath("v", arg0ElemType.ref());
+                    DBSPExpression cast = var.deref().cast(elemType).closure(var.asParameter());
+                    arg0 = new DBSPApplyExpression("map", type, arg0.borrow(), cast);
+                }
+
+                String method = "array_append";
+                if (arg0type.mayBeNull)
+                    method += "N";
+
+                return new DBSPApplyExpression(node, method, type, arg0, arg1);
```
Comment (abhizer) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java:975]: We pass by value instead of borrowing here. This is so that for `Option<Vec<T>>` borrowing would pass `&Option<Vec<T>>` instead of  a `&[T]`.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/ArrayTests.java [hunk 5]
```diff
@@ -289,4 +289,122 @@ public void testElementNull() {
...
+                                new DBSPI32Literal(7),
+                                new DBSPI32Literal(4))
+                )
+        );
+
+        this.testQuery(ddl, "SELECT ARRAY_APPEND(val, 4) FROM ARR_TBL", new InputOutputPair(input, result));
+    }
+
+    @Test @Ignore
+    public void testArrayAppendInnerNullable() {
```
Comment (abhizer) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/ArrayTests.java:370]: This test fails because Calcite seems to expect a `Vec<T>` instead of `Vec<Option<T>>` for even when the input array has nullable elements. 


```rust
error[E0308]: mismatched types
    --> src/lib.rs:172:44
     |
172  |     append_to_collection_handle(&input0.0, &streams.0);
     |     ---------------------------            ^^^^^^^^^^ expected `std::option::Option<i32>`, found `i32`
     |     |
     |     arguments to this function are incorrect
     |
     = note: expected reference `&CollectionHandle<dbsp::utils::Tup1<std::option::Option<Vec<std::option::Option<i32>>>>, i64>`
                found reference `&CollectionHandle<dbsp::utils::Tup1<std::option::Option<Vec<i32>>>, i64>`
note: function defined here
    --> /home/abhizer/feldera/sql-to-dbsp-compiler/lib/sqllib/src/lib.rs:1427:8
     |
1427 | pub fn append_to_collection_handle<K>(data: &WSet<K>, handle: &CollectionHandle<K, Weight>)
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/lib.rs:174:58
    |
174 |       assert!(must_equal(&read_output_handle(&streams.1), &zset!(
    |  __________________________________________________________^
175 | |         Tup1::new(Some(vec!(
176 | |             None::<i32>, Some(1i32), Some(2i32), Some(3i32), Some(4i32), ))) => 1,
177 | |         Tup1::new(Some(vec!(
178 | |             None::<i32>, Some(5i32), Some(6i32), Some(7i32), Some(4i32), ))) => 1,
179 | |     )), r#"
    | |_____^ expected `i32`, found `std::option::Option<i32>`
    |
    = note: expected struct `OrdZSet<dbsp::utils::Tup1<std::option::Option<Vec<i32>>>, i64>`
               found struct `OrdZSet<dbsp::utils::Tup1<std::option::Option<Vec<std::option::Option<i32>>>>, {integer}>`
    = note: this error originates in the macro `zset` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `temp` (lib test) due to 2 previous errors
```

Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/ArrayTests.java:370]: Yes, you found a new bug in Calcite: https://issues.apache.org/jira/browse/CALCITE-6275
I submitted a PR, if they don't like it we can change the parser unilaterally to do the right thing here.
